### DISCUSSION TITLE
RAG over multiple documents

### DISCUSSION THREAD

**Discussion by rosmur** [open-webui/open-webui#995] [Q&A]
Hello, As far as I can tell OpenWebUI allows to do RAG on only a single document? Is there a way to do RAG for a whole folder of docs?

**Comment by andykaufseo** [comment]
Looking for the same thing, even if i add 100 docs to the docs folder, it only uses the first 5-10 of them for rag and ignores the other ones...

**Reply by pwab** [reply]
Is this a problem related to scanning the documents folder?
![image](https://github.com/open-webui/open-webui/assets/19271185/a2022833-a5de-4fd7-b2bb-6bce05e93f61)

Or did you somehow manage to add 100 documents to the conversations context?
I'm not sure how open-webui is storing the information of the embedded documents and how they are added to the context but it could be an issue with context length. And as far as I know the context length is depending on the used base model and its parameters.

**Reply by justinh-rahb** [reply]
It is ***definitely*** a context-length issue if you've got that many documents. Your best chance would be using the new 128k Command-R model out now, assuming you have enough VRAM for that. Otherwise next best is probably 32k with Mistral/Mixtral models.

**Reply by raoulg** [reply]
In principle RAG should allow you to _potentially_ query all documents. It's just that not all documents are relevant. Depending on your question, you get a relevant top k of documents. A lot of times, you won't need more than k documents to formulate an answer.

In the case you actually do need _all_ documents for a question, knowledge graphs (KG) might be of help. I am currently looking into something like autoKG, that makes a KG (think: mindmap) of your documents. Might be interesting to add that.

**Comment by pwab** [comment]
Well you can add multiple documents to the context of a single conversation with using the `#` multiple times. As far as I know there is currently no way to reference the whole set of embedded documents or a subfolder.

This maybe could be solved with building a modelfile first that references a set of documents. So somehow related to #529.

**Reply by ghost** [reply]
Actually you can use tags to mark a subset of documents:
![image](https://github.com/open-webui/open-webui/assets/126992880/0fa8104b-4811-4755-9cf5-890300e71a27)
in addition to the "all documents" conversation option:
![image](https://github.com/open-webui/open-webui/assets/126992880/42a8d1e3-1ba4-4a4a-9676-d194468645ba)

**Reply by pwab** [reply]
That is awesome! I have to try this asap.
